File: 2. Add Two Numbers.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
struct ListNode *addTwoNumbers(struct ListNode* l1, struct ListNode* l2) {
    struct ListNode *head, *prev;
    int sum, carry;

    head = l1;
    carry = 0;
    while (l1 && l2) {
        sum = l1->val + l2->val + carry;
        l1->val = sum % 10;
        carry = sum / 10;
        prev = l1;
        l1 = l1->next;
        l2 = l2->next;
    }
    l1 = l1 ? l1 : l2;
    prev->next = l1;
    while (l1) {
        prev = l1;
        sum = l1->val + carry;
        l1->val = sum % 10;
        carry = sum / 10;
        l1 = l1->next;
    }
    if (carry == 1) {
        struct ListNode *p;
        p = malloc(sizeof(struct ListNode));
        p->val = carry;
        p->next = NULL;

        prev->next = p;
    }
    return head;
}

File: Python3/2. Add Two Numbers.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def addTwoNumbers(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        dummy = ListNode(0)
        prev = dummy
        carry = 0
        while l1 or l2:
            x = l1.val if l1 else 0
            y = l2.val if l2 else 0
            summary = x + y + carry

            cur, carry = ListNode(summary % 10), summary // 10
            if l1:
                l1 = l1.next
            if l2:
                l2 = l2.next
            prev.next, prev = cur, cur

        if carry:
            cur = ListNode(carry)
            prev.next = cur

        return dummy.next
